### ISSUE TITLE
Amazon Linux 2 - error on running "make clusterawsadm": "- bash: - : invalid option"

### ISSUE THREAD

**Issue by chunywan** [kubernetes-sigs/cluster-api-provider-aws#2804]
/kind bug

**What steps did you take and what happened:**
The following error happens when running `make clusterawsadm`
```
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
go build -ldflags "" -o bin/clusterawsadm ./cmd/clusterawsadm
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
make: *** [clusterawsadm] Error 1
```

With debug info (make clusterawsadm -d):
```
GNU Make 3.82
Built for x86_64-koji-linux-gnu
Copyright (C) 2010  Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Reading makefile `Makefile'...
Reading makefile `common.mk' (search path) (no ~ expansion)...
Reading makefile `versions.mk' (search path) (no ~ expansion)...
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
Updating makefiles....
 Considering target file `versions.mk'.
...
...
Updating goal targets....
Considering target file `clusterawsadm'.
 File `clusterawsadm' does not exist.
 Finished prerequisites of target file `clusterawsadm'.
Must remake target `clusterawsadm'.
Invoking recipe from Makefile:223 to update target `clusterawsadm'.
go build -ldflags "" -o bin/clusterawsadm ./cmd/clusterawsadm
bash: - : invalid option
bash: - : invalid option
bash: - : invalid option
Putting child 0x8a8750 (clusterawsadm) PID 578527 on the chain.
Live child 0x8a8750 (clusterawsadm) PID 578527
bash: - : invalid option
Reaping losing child 0x8a8750 PID 578527
make: *** [clusterawsadm] Error 1
Removing child 0x8a8750 PID 578527 from chain.
``` 

**What did you expect to happen:**
I expect the binary `clusterawsadm` can be built successfully.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
It works well on my MacOS, but failed on Amazon Linux 2.

**Environment:**

- Cluster-api-provider-aws version:  v0.7.0
- Kubernetes version: (use `kubectl version`): v1.22.1
- OS (e.g. from `/etc/os-release`): 
```
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2:-:internal"
HOME_URL="https://amazonlinux.com/"
VARIANT="internal"
```

**Comment by chunywan** [comment]
Mostly this is related to the conflicts between `ONESHELL` special target and `shell function $(shell)`. With the following simplified `Makefile`, it still get error `bash: - : invalid option`:
```Makefile
# Ensure Make is run with bash shell as some syntax below is bash-specific
SHELL=bash
.ONESHELL:
.SHELLFLAGS := -eu -o pipefail -c
UID := $(shell id -u)
``` 

From my testing, it works well with `make-4.0`, `make-4.1`, `make-4.2` and `make-4.3`. Only make-3.82 has this issue.

In summary, we can upgrade make to 4.x or downgrade to 3.81 to run `make clusterawsadm`.

**Comment by chunywan** [comment]
This is the link to the bug found in `make-3.82` and  fixed in `make-4.0` release. It confirms my findings above.
[bug #35397: make bug with ONESHELL and SHELLFLAGS](http://savannah.gnu.org/bugs/index.php?35397)

So I will close this github issue.

**Comment by randomvariable** [comment]
Just to clarify, was the problem building on MacOS or on AML 2?

On MacOS, I mostly use make from Homebrew and assume the standard one isn't going to work. What does Homebrew ship with at the mo?

And is the issue that AML2 ships with make 3.82 RPMs? Are make-4.0, make-4.1, make-4.2 and make-4.3 different packages in the yum repos for AML2?
